Don't wait for sync to update readiness

timstclair opened the pull request:
Push status updates as soon as readiness state changes for containers,
rather than waiting for the sync loop to update the status. In
particular, this should help new containers to come online faster.

Additionally, consolidates prober test helpers into a single file.

ashcrow commented:
Shippable error looks totally unrelated:

```
--- /tmp/tmp.SAVsU8E2pr/docs/man/man1/kubectl-autoscale.1	2015-10-21 01:35:17.961069431 +0000
+++ /root/workspace/src/k8s.io/kubernetes/docs/man/man1/kubectl-autoscale.1	2015-10-21 00:33:46.029069431 +0000
@@ -72,7 +72,7 @@ An autoscaler can automatically increase
     If non\-empty, sort list types using this field specification.  The field specification is expressed as a JSONPath expression (e.g. 'ObjectMeta.Name'). The field in the API resource specified by this JSONPath expression must be an integer or a string.
 
 .PP
-\fB\-\-template\fP=""
+\fB\-t\fP, \fB\-\-template\fP=""
     Template string or path to template file to use when \-o=go\-template, \-o=go\-template\-file. The template format is golang templates [
 \[la]http://golang.org/pkg/text/template/#pkg-overview\[ra]].
 
Generated docs are out of date. Please run hack/update-generated-docs.sh
!!! Error in ./hack/verify-generated-docs.sh:28
```

yujuhong commented on pkg/kubelet/prober/manager_test.go:
```diff
@@ -300,3 +305,27 @@ func waitForWorkerExit(m *manager, workerPaths []probeKey) error {
+	condition := func() (bool, error) {
+		status, ok := m.statusManager.GetPodStatus(testPodUID)
+		if !ok {
+			return false, fmt.Errorf("Status not found: %q", testPodUID)
```
The convention for golang error message is to start with lower case letters.

timstclair replied:
Done.

yujuhong commented:
Hmm...I need to think about this more because it may break mirror pods. We update the status of the mirror pod instead if such a pod exists at the end of `syncPod` https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1350. Perhaps we should push the static->mirror pod translation to status manager to avoid this whole shenanigans, but that means status manager needs to query the pod manager for information...

Sorry for not catching this earlier.  :-(

timstclair commented:
Fortunately the update model is pretty flexible, and the readiness update can easily be moved into the kubelet (I'd prefer that over passing the podmanager to the probe.Manager).

yujuhong commented:
> Fortunately the update model is pretty flexible, and the readiness update can easily be moved into the kubelet (I'd prefer that over passing the podmanager to the probe.Manager).

Oh I didn't not suggest prober to access the pod manager. All api-facing components would need to do the translation, so maybe it's *not too bad* to pass pod manager to status manager. 

Where in the kubelet do you want to move the readiness update too?

timstclair commented:
To elaborate on my proposal to move the updating to kubelet:

- Initialize readiness manager in kubelet (pass to probe.Manager, we were doing this before)
- move update readiness function to kubelet, call it the same way (util.Forever) somewhere in kubelet (probably Run)
- In update readiness, do:
```go
...
if !kubepod.IsMirrorPod(pod) {
  pod = kl.podManager.GetMirrorPodByPod(pod)
}
...
```

I'm still a little fuzzy on what the mirror pod does, so if you think it's better to move the logic to status manager, I'm happy to go with that solution. Do we use the mirror pod for anything other than status updates?

timstclair commented:
Rebased & updated with changes to `status.Manager`

PTAL

timstclair commented:
Fixed broken imports.

yujuhong commented:
Nice cleanup! LGTM

timstclair commented:
@k8s-bot test this

wojtek-t commented:
@timstclair can you please rebase?

wojtek-t commented:
@timstclair - please rebase

timstclair commented:
Rebased.

wojtek-t commented:
@timstclair - seems that hack/verify-gofmt failed...

tallclair committed:
Don't wait for sync to update readiness

Push status updates as soon as readiness state changes for containers,
rather than waiting for the sync loop to update the status. In
particular, this should help new containers to come online faster.

Additionally, consolidates prober test helpers into a single file.

timstclair commented:
Fixed. I need to fix my laptop dev environment...

timstclair commented:
@k8s-bot test this

wojtek-t commented:
@k8s-bot test this please
